### PR TITLE
[93] Refactor General View diagram palette

### PR DESCRIPTION
**Pull request by jerome-obeo**
Adding 3 incomplete sections:
 - Structure
 - Interconnection
 - Extension

### CODE REVIEW

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/DefinitionNodeDescriptionProvider.java [hunk 1]
```diff
@@ -73,14 +73,11 @@ public void link(DiagramDescription diagramDescription, IViewDiagramElementFinde
...
 
-        GeneralViewDiagramDescriptionProvider.DEFINITIONS.forEach(definition -> {
-            var optNodeDescription = cache.getNodeDescription(GVDescriptionNameGenerator.getNodeName(definition));
-            allTargetNodeDescriptions.add(optNodeDescription.get());
-        });
-
-        GeneralViewDiagramDescriptionProvider.USAGES.forEach(usage -> {
-            var optNodeDescription = cache.getNodeDescription(GVDescriptionNameGenerator.getNodeName(usage));
-            allTargetNodeDescriptions.add(optNodeDescription.get());
+        GeneralViewDiagramDescriptionProvider.TOOL_SECTIONS.forEach((sectionName, elements) -> {
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/DefinitionNodeDescriptionProvider.java:76]: Why did you use TOOL_SECTIONS here? It seems strange to use the TOOL_SECTIONS to compute the node descriptions list that will be used by this Definition Node Palette.

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/GeneralViewDiagramDescriptionProvider.java [hunk 2]
```diff
@@ -86,11 +86,36 @@ public class GeneralViewDiagramDescriptionProvider implements IRepresentationDes
...
+                    SysmlPackage.eINSTANCE.getPartDefinition()
+                    )),
+            Map.entry("Interconnection", List.of(
+                    SysmlPackage.eINSTANCE.getInterfaceUsage(), 
+                    SysmlPackage.eINSTANCE.getInterfaceDefinition(), 
+                    SysmlPackage.eINSTANCE.getPortUsage(), 
+                    SysmlPackage.eINSTANCE.getPortDefinition()
+                    )),
+            Map.entry("Extension", List.of(
+                    SysmlPackage.eINSTANCE.getMetadataUsage(),
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/GeneralViewDiagramDescriptionProvider.java:115]: If you add MetadataUsage, you should also add it in the USAGE list above.
Furthermore, are you sure to fully handled it in this PR? MetadataUsage has a specific graphical notation, which doesn't seem to be handled here.

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/PackageNodeDescriptionProvider.java [hunk 3]
```diff
@@ -169,12 +162,11 @@ private DropNodeTool createDropFromDiagramTool(IViewDiagramElementFinder cache)
...
-        GeneralViewDiagramDescriptionProvider.USAGES.forEach(usage -> {
-            nodeTools.add(this.createNodeTool(cache.getNodeDescription(GVDescriptionNameGenerator.getNodeName(usage)).get(), usage));
+        GeneralViewDiagramDescriptionProvider.TOOL_SECTIONS.forEach((sectionName, elements) -> {
+            elements.forEach(element -> {
+                var optNodeDescription = cache.getNodeDescription(GVDescriptionNameGenerator.getNodeName(element));
+                nodeTools.add(this.createNodeTool(optNodeDescription.get(), element));                
+            });
         });
 
         nodeTools.add(this.createNodeTool(cache.getNodeDescription(PackageNodeDescriptionProvider.NAME).get(), SysmlPackage.eINSTANCE.getPackage()));
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/PackageNodeDescriptionProvider.java:172]: You use TOOL_SECTIONS above to create add the tools needed but forget to remove this line.
But Package is already handled by the TOOL_SECTIONS.

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/UsageNodeDescriptionProvider.java [hunk 4]
```diff
@@ -73,16 +73,13 @@ public void link(DiagramDescription diagramDescription, IViewDiagramElementFinde
 
         var allTargetNodeDescriptions = new ArrayList<NodeDescription>();
 
-        GeneralViewDiagramDescriptionProvider.DEFINITIONS.forEach(definition -> {
-            var optNodeDescription = cache.getNodeDescription(GVDescriptionNameGenerator.getNodeName(definition));
-            allTargetNodeDescriptions.add(optNodeDescription.get());
+        GeneralViewDiagramDescriptionProvider.TOOL_SECTIONS.forEach((sectionName, elements) -> {
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/UsageNodeDescriptionProvider.java:76]: Same remark than for the DefinitionNodeDescriptionProvider.

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/EmptyDiagramNodeDescriptionProvider.java [hunk 5]
```diff
@@ -177,4 +160,31 @@ private NodeTool addExistingElementsTool() {
...
+        
+        sections.add(addElementsToolSection(cache));
+        
+        return sections.toArray(NodeToolSection[]::new);
+    }
+    
+    private NodeTool[] createElementsOfToolSection(IViewDiagramElementFinder cache, List<EClass> elements) {
+        var nodeTools = new ArrayList<NodeTool>();
+
+        elements.forEach(definition -> {
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/EmptyDiagramNodeDescriptionProvider.java:182]: Could you rename this `definition` variable please? I don't think is named correctly.

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/EmptyDiagramNodeDescriptionProvider.java [hunk 6]
```diff
@@ -177,4 +160,31 @@ private NodeTool addExistingElementsTool() {
...
+        var sections = new ArrayList<NodeToolSection>();
+        
+        GeneralViewDiagramDescriptionProvider.TOOL_SECTIONS.forEach((sectionName, elements) -> {
+            NodeToolSectionBuilder sectionBuilder = this.diagramBuilderHelper.newNodeToolSection()
+                    .name(sectionName)
+                    .nodeTools(this.createElementsOfToolSection(cache, elements));
+            sections.add(sectionBuilder.build());
+        });
+        
+        sections.add(addElementsToolSection(cache));
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/EmptyDiagramNodeDescriptionProvider.java:174]: Could you add `this.` in front of every class variables/methods please? Could you also check everywhere else you have modified code in this PR?

**Review by AxelRICHARD — CHANGES_REQUESTED** [review]

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/GeneralViewDiagramDescriptionProvider.java [hunk 7]
```diff
@@ -86,11 +86,36 @@ public class GeneralViewDiagramDescriptionProvider implements IRepresentationDes
...
             Map.entry(SysmlPackage.eINSTANCE.getItemUsage(),             List.of(SysmlPackage.eINSTANCE.getUsage_NestedAttribute(), SysmlPackage.eINSTANCE.getUsage_NestedReference())),
             Map.entry(SysmlPackage.eINSTANCE.getPartUsage(),             List.of(SysmlPackage.eINSTANCE.getUsage_NestedAttribute(), SysmlPackage.eINSTANCE.getUsage_NestedPort())),
             Map.entry(SysmlPackage.eINSTANCE.getPortUsage(),             List.of(SysmlPackage.eINSTANCE.getUsage_NestedAttribute(), SysmlPackage.eINSTANCE.getUsage_NestedReference()))
             );
+    
+    public static final Map<String, List<EClass>> TOOL_SECTIONS = Map.ofEntries(
+            Map.entry("Structure", List.of(
+                    SysmlPackage.eINSTANCE.getAttributeUsage(),
+                    SysmlPackage.eINSTANCE.getAttributeDefinition(),
+                    SysmlPackage.eINSTANCE.getEnumerationUsage(),
```
[the remark this answers is not in the document]
Comment (jerome-obeo) [reply backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/GeneralViewDiagramDescriptionProvider.java:100]: I followed the Figma file Melanie is working on describing each palette section. I can see entries for Enumeration Usage and Enumeration Definition concepts in the Structure section. It seems that she has not yet separated node tools and diagram tools.
I will remove it from the diagram tools.

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/GeneralViewDiagramDescriptionProvider.java [hunk 8]
```diff
@@ -86,11 +86,36 @@ public class GeneralViewDiagramDescriptionProvider implements IRepresentationDes
             Map.entry(SysmlPackage.eINSTANCE.getPartDefinition(),        List.of(SysmlPackage.eINSTANCE.getDefinition_OwnedAttribute(), SysmlPackage.eINSTANCE.getDefinition_OwnedPort())),
             Map.entry(SysmlPackage.eINSTANCE.getPortDefinition(),        List.of(SysmlPackage.eINSTANCE.getDefinition_OwnedAttribute(), SysmlPackage.eINSTANCE.getDefinition_OwnedPort(), SysmlPackage.eINSTANCE.getDefinition_OwnedReference())),
             Map.entry(SysmlPackage.eINSTANCE.getAttributeUsage(),        List.of(SysmlPackage.eINSTANCE.getUsage_NestedAttribute(), SysmlPackage.eINSTANCE.getUsage_NestedReference())),
+            Map.entry(SysmlPackage.eINSTANCE.getEnumerationUsage(),      List.of(SysmlPackage.eINSTANCE.getEnumerationUsage_EnumerationDefinition())),
```
[the remark this answers is not in the document]
Comment (jerome-obeo) [reply backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/GeneralViewDiagramDescriptionProvider.java:89]: Agree

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/PackageNodeDescriptionProvider.java (hunk 3 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/PackageNodeDescriptionProvider.java:172]: Done

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/edges/DependencyEdgeDescriptionProvider.java [hunk 9]
```diff
@@ -68,16 +68,12 @@ public void link(DiagramDescription diagramDescription, IViewDiagramElementFinde
...
-            var optNodeDescription = cache.getNodeDescription(GVDescriptionNameGenerator.getNodeName(definition));
-            edgeDescription.getSourceNodeDescriptions().add(optNodeDescription.get());
-            edgeDescription.getTargetNodeDescriptions().add(optNodeDescription.get());
-        });
-
-        GeneralViewDiagramDescriptionProvider.USAGES.forEach(usage -> {
-            var optNodeDescription = cache.getNodeDescription(GVDescriptionNameGenerator.getNodeName(usage));
-            edgeDescription.getSourceNodeDescriptions().add(optNodeDescription.get());
-            edgeDescription.getTargetNodeDescriptions().add(optNodeDescription.get());
+        GeneralViewDiagramDescriptionProvider.TOOL_SECTIONS.forEach((sectionName, elements) -> {
```
[the remark this answers is not in the document]
Comment (jerome-obeo) [reply backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/edges/DependencyEdgeDescriptionProvider.java:71]: Initially, my goal was to replace the USAGES and DEFINITIONS with TOOL_SECTIONS, but following the single responsibility principle, TOOL_SECTIONS should only deal with tool sections.
So, I restored the previous code.

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/DefinitionNodeDescriptionProvider.java (hunk 1 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/DefinitionNodeDescriptionProvider.java:76]: You are right I reverted this change

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/UsageNodeDescriptionProvider.java (hunk 4 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/UsageNodeDescriptionProvider.java:76]: Reverted

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/EmptyDiagramNodeDescriptionProvider.java (hunk 5 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/EmptyDiagramNodeDescriptionProvider.java:182]: Changed

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/EmptyDiagramNodeDescriptionProvider.java (hunk 6 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/EmptyDiagramNodeDescriptionProvider.java:174]: Done

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/GeneralViewDiagramDescriptionProvider.java (hunk 2 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/GeneralViewDiagramDescriptionProvider.java:115]: Agree. I've removed `MetadataUsage` for this PR.

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/GeneralViewNodeDescriptionProviderSwitch.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+import org.eclipse.syson.util.SysmlEClassSwitch;
+
+/**
+ * Switch retrieving the list of NodeDescriptionProvider for each SysMLv2 concept represented in the General View diagram.
+ * 
+ * @author Jerome Gout
+ */
+public class GeneralViewNodeDescriptionProviderSwitch extends SysmlEClassSwitch<INodeDescriptionProvider> {
+    
+    private IColorProvider colorProvider;
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/GeneralViewNodeDescriptionProviderSwitch.java:44]: Could you please add `final` to this field please?

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/GeneralViewNodeDescriptionProviderSwitch.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+import org.eclipse.syson.sysml.PortUsage;
+import org.eclipse.syson.sysml.SysmlPackage;
+import org.eclipse.syson.util.SysmlEClassSwitch;
+
+/**
+ * Switch retrieving the list of NodeDescriptionProvider for each SysMLv2 concept represented in the General View diagram.
+ * 
+ * @author Jerome Gout
+ */
+public class GeneralViewNodeDescriptionProviderSwitch extends SysmlEClassSwitch<INodeDescriptionProvider> {
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/GeneralViewNodeDescriptionProviderSwitch.java:42]: Sorry I forgot to tell you on the first commit, could you please add all the `caseXXX` in alphabetical order please?

**Review by AxelRICHARD — CHANGES_REQUESTED** [review]

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/GeneralViewNodeDescriptionProviderSwitch.java (hunk 10 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/GeneralViewNodeDescriptionProviderSwitch.java:44]: Done.

File: backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/GeneralViewNodeDescriptionProviderSwitch.java (hunk 11 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-general-view/src/main/java/org/eclipse/syson/diagram/general/view/nodes/GeneralViewNodeDescriptionProviderSwitch.java:42]: Done.

**Review by AxelRICHARD — COMMENTED** [review]
Thank you for your work on this PR.
Just please squash the commits, fix the copyright in MetadataUsageItemProvider and it should be ok.

**Review by AxelRICHARD — APPROVED** [review]

**Pull request merged**